beacon_node/store/src/hot_cold_store.rs

```diff
@@ -798,21 +822,41 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> HotColdDB<E, Hot, Cold>
         Ok(key_value_batch)
     }
 
-    pub fn do_atomically(&self, batch: Vec<StoreOp<E>>) -> Result<(), Error> {
-        // Update the block cache whilst holding a lock, to ensure that the cache updates atomically
-        // with the database.
+    pub fn do_atomically_with_block_and_blobs_cache(
+        &self,
+        batch: Vec<StoreOp<E>>,
+    ) -> Result<(), Error> {
+        let mut hot_db_cache_ops = Vec::new();
+
+        let (blobs_ops, hot_db_ops) = batch.into_iter().partition(|store_op| match store_op {
+            StoreOp::PutBlobs(_, _) | StoreOp::DeleteBlobs(_) => true,
+            StoreOp::PutBlock(_, _) | StoreOp::DeleteBlock(_) => {
+                hot_db_cache_ops.push(store_op.clone());
+                false
+            }
+            _ => false,
+        });
+
+        // Update database whilst holding a lock on cache, to ensure that the cache updates
+        // atomically with the database.
         let mut guard = self.block_cache.lock();
         let mut guard_blob = self.blob_cache.lock();
 
-        for op in &batch {
+        self.hot_db
+            .do_atomically(self.convert_to_kv_batch(hot_db_ops)?)?;
+
+        let blob_cache_ops = blobs_ops.clone();
+        let blobs_db = self.blobs_db.as_ref().unwrap_or(&self.cold_db);
+        // todo(emhane): do we want to restore the hot db writes if this fails?
```

Comment:
Are you asking if we should delete the blocks from the blob DB if the write fails? It seems like that would be the safest thing to do. However, I see that `do_atomically_with_block_and_blobs_cache` might also process other operations and reverting them all might get a little complex.

I would be tempted to try writing to the `blobs_db` first, then doing the `hot_db`. If the second write fails then I imagine it would be simpler to try and remove the blobs rather than revert the `hot_db` calls.

Perhaps @michaelsproul has some opinions on this DB atomicity decision?

Reply:
Yeah, that's exactly what I mean. Ok, that's a good idea to reverse the order and delete blobs instead.